pkg/controller/util/endpoint/rate_limiting_queue.go

```diff
@@ -0,0 +1,114 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package endpoint
+
+import (
+	"sync"
+
+	"k8s.io/client-go/util/workqueue"
+)
+
+// RateLimitingQueue implements a wrapper on workqueue.RateLimitingInterface that also rate limits initial key additions
+// to the queue (unlike workqueue.RateLimitingInterface which rate limits only error retries).
+// TODO(mborsz): Rename to something that doesn't collide with workqueue.RateLimitingInterface (or refactor workqueue.RateLimitingInterface).
```

Comment:
I think we should do that now (before merging this PR).

Reply:
I don't have good suggestion though, maybe:

DeduplicatingRateLimittingQueue ?

Reply:
DOne